DOCSP-30561 Performance Considerations (#30)
mongodb/docs-rust: source/fundamentals/performance.txt

   :class: singlecol

Overview
--------

In this guide, you can learn about best practices for optimizing performance of the {+driver-short+}.
The ``Client`` handles many aspects of database connection behind the scenes that can require manual management for other database drivers;
it discovers server topology, monitors it for any changes, and maintains an internal connection pool.
This has implications for how a ``Client`` should be used for best performance.

.. _rust-performance-client-lifecycle:

Best Practices
--------------

Client Lifecycle
~~~~~~~~~~~~~~~~

Creating a new ``Client`` instance is relatively slow and resource-intensive.
Therefore a ``Client`` instance should be as long-lived as possible.
The ``Client`` type does its own pooling internally.
Attempting to maintain a pool of ``Client`` instances will result in worse performance than using a single instance.

The ``Client`` type is implemented using an internal ``Arc``,
it can safely be shared across threads or tasks.
Cloning it to pass to new contexts is extremely cheap.
This means you can create a ``Client`` instance once at application startup and continually reuse it.

Consider the following code.

> It constructs and tears down a ``Client`` instance with every request.

Reviewer: S: Improve the introduction of this code example by actually describing what it contains at a more specific level:
Something like:

"Creating a new ``Client`` instance for each request results in slower performance. The following code defines a method that instantiates a new ``Client`` before performing other operations:"

However, I'm not sure it serves users to include content on this page that describes what NOT to do. It might be confusing and a user quickly reading the page might copy the wrong snippet. Happy to talk more about this, but my suggestion is to only describe the downsides of creating a new client but only show the code snippet for the more efficient best practice code.

.. literalinclude:: /includes/fundamentals/code-snippets/performance.rs
   :language: rust
   :dedent:
   :start-after: start-perf-client-slow
   :end-before: end-perf-client-slow

Instead, pass a longer-lived ``Client`` instance to the handler function.
The following code accepts a pointer to an existing ``Client`` instance,

> allowing many requests to share the same client:

Reviewer: S: same comment as above, improve this code introduction

.. literalinclude:: /includes/fundamentals/code-snippets/performance.rs
   :language: rust
   :dedent:
   :start-after: start-perf-client-faster
   :end-before: end-perf-client-faster

.. _rust-performance-runtime:

Runtime
~~~~~~~

> A ``Client`` instance is bound to the instance of the ``tokio`` or ``async-std`` runtime in which it was created.

Reviewer: S; avoid passive voice

```suggestion
A ``Client`` instance is bound to the instance of the ``tokio`` or ``async-std`` runtime in which you created it.
```

> Attempting to execute operations on a different runtime instance can cause incorrect behavior and unpredictable failures.

Reviewer: ```suggestion
If you use a ``Client`` instance to execute operations on a different runtime, you might experience unexpected behavior or failures.
```

> This is easy to accidentally invoke when testing,
> as the ``tokio::test`` or ``async_std::test`` helper macros create a new runtime for each test.

Reviewer: ```suggestion
If you are using the ``test`` helper macro from the ``tokio`` or ``async_std`` crate to test your application, you might accidentally run operations on a different runtime than you intended. This is because these helper macros 
create a new runtime for each test.
```

.. literalinclude:: /includes/fundamentals/code-snippets/performance-new-runtime.rs
   :language: rust
   :dedent:

> 

Reviewer: I: This code needs an introduction that explains why it will fail, etc.
S: describe the code before including it

> To work around this issue, either create a new ``Client`` instance for every ``async`` test,
> or bundle the runtime along with the ``Client`` instance and don't use the test helper macros.

Reviewer: ```suggestion
To avoid this issue, you can take the following approaches:
- Create a new ``Client`` instance for every ``async`` test
- Attach the runtime to the ``Client`` instance without using the ``test`` helper macros
```

.. literalinclude:: /includes/fundamentals/code-snippets/performance-bundle-runtime.rs
   :language: rust
   :dedent:

or

.. literalinclude:: /includes/fundamentals/code-snippets/performance-new-client.rs
   :language: rust

>    :dedent:

Reviewer: I: code snippets need introductions
S: add introductions for each

.. _rust-performance-parallelism:

Parallelism
~~~~~~~~~~~

> Where data operations are naturally parallelizable,
> spawning many asynchronous tasks that use the driver concurrently is often the best way to achieve maximum performance,
> as the driver is designed to work well in such situations.

Reviewer: ```suggestion
If you can run parallel data operations, creating asynchronous, concurrent tasks achieves maximum performance.
```

.. literalinclude:: /includes/fundamentals/code-snippets/performance-parallel.rs
   :language: rust
   :dedent:

> 

Reviewer: I: needs introduction
S: add intro "The following code xxx"

Additional Information
----------------------

For more information about the concepts in this guide, see the following documentation:

- `Client() <{+api+}/struct.Client.html>`__
- `sync <{+api+}/sync/index.html>`__

> - `Arc <https://doc.rust-lang.org/std/sync/struct.Arc.html>`__

Reviewer: S: add a heading for API documentation

.. _rust-performance:

==========================
Performance Considerations
==========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to optimize performance of the {+driver-short+}.
To connect to MongoDB, you must create a ``Client`` instance. Your ``Client``
instance automatically handles most aspects of connection, such as discovering server topology, monitoring
your connection, and maintaining an internal connection pool. This guide describes best practices to configure and use your ``Client`` instance.

.. _rust-performance-client-lifecycle:

Client Lifecycle
----------------

We recommend that you reuse your client across sessions and operations.

> You can use the same ``Client`` instance to perform multiple tasks, instead of creating a new one each time.
> The ``Client`` type is safe for concurrent use by multiple threads.

Reviewer: S; combine sentences
```suggestion
You can use the same ``Client`` instance to perform multiple tasks, as the ``Client`` type is safe for concurrent use by multiple threads.
```

Creating a new ``Client`` instance for each request results in slower performance.

The following code creates a method that accepts a pointer to an existing ``Client`` instance,

> which allows you to execute many requests by using same client:

Reviewer: S: style guidance around "execute": https://www.mongodb.com/docs/meta/style-guide/terminology/terms-for-global-audience/#:~:text=Windows%20is%20discontinued.-,execute,-v

Also replace elsewhere in the page

```suggestion
which allows you to perform many operations by using same client:
```

.. literalinclude:: /includes/fundamentals/code-snippets/performance.rs
   :language: rust
   :dedent:
   :start-after: start-perf-client-faster
   :end-before: end-perf-client-faster

.. _rust-performance-runtime:

> Runtime

Reviewer 2: A similar `Runtime` section is included in the [Performance - MongoDB Rust Driver](https://mongodb.github.io/mongo-rust-driver/manual/performance.html#runtime) page. I expect this section is less valuable to new users of the driver, as this is intended for testing.

Consider removing the `Runtime` section. Alternatively: move it below the `Parallelism` section. I expect the `Parallelism` is more relevant to users.

-------

A ``Client`` instance is bound to the instance of the ``tokio`` or ``async-std`` runtime in which you created it.
If you use a ``Client`` instance to execute operations on a different runtime, you might experience unexpected behavior or failures.

If you are using the ``test`` helper macro from the ``tokio`` or ``async_std`` crate to test your application,
you might accidentally run operations on a different runtime than you intended.
This is because these helper macros create a new runtime for each test.
You can use one of the following strategies to avoid this issue:

- Attach the runtime to the ``Client`` instance without using the ``test`` helper macros.
- Create a new ``Client`` instance for every ``async`` test.

This example follows the first strategy and creates a global runtime used only for testing.
In the following code, the ``test_list_dbs()`` method uses a client that manually connects to this runtime
to list databases in the deployment.

.. literalinclude:: /includes/fundamentals/code-snippets/performance-bundle-runtime.rs
   :language: rust
   :dedent:

> As an example of the latter,

Reviewer: ```suggestion
Implementing the second strategy,
```

> the following code creates a new ``Client`` instance for each test run with ``tokio::test``.

Reviewer: ```suggestion
the following code creates a new ``Client`` instance for each test run by using the ``tokio::test`` macro.
```

> As an example of the latter,
> the following code creates a new ``Client`` instance for each test run with ``tokio::test``.
> This ensures there is no unintended interaction between runtimes.

Reviewer: ```suggestion
As an example of the latter,
the following code creates a new ``Client`` instance for each test run with ``tokio::test``, 
ensuring that there is no unintended interaction between runtimes:
```

.. literalinclude:: /includes/fundamentals/code-snippets/performance-new-client.rs
   :language: rust
   :dedent:

.. _rust-performance-parallelism:

Parallelism
-----------

> If you can run parallel data operations, creating asynchronous, concurrent tasks achieves maximum performance.

Reviewer: ```suggestion
If you can run parallel data operations, you can optimize performance by running asynchronous, concurrent tasks.
```

The following code uses the ``task`` module from the ``tokio`` crate to create separate, concurrent
tasks for multiple data operations:

.. literalinclude:: /includes/fundamentals/code-snippets/performance-parallel.rs
   :language: rust
   :dedent:

API Documentation
-----------------

- `Client() <{+api+}/struct.Client.html>`__
- `sync <{+api+}/sync/index.html>`__

> .. TODO link to Async page when done

Reviewer: S: structure this section like the other fundamentals pages by having the Additional Information heading with links, and then the API Documentation subheading. Under the addtl information heading. you could link to the runtimes page, https://preview-mongodbrustagir.gatsbyjs.io/rust/DOCSP-30545-serialization/fundamentals/connections/connection-guide/#mongodb-client,
